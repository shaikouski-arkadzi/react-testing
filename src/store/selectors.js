export const selectorTodos = (state) => state.todos;
