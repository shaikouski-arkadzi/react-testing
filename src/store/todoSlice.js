import { createSlice } from "@reduxjs/toolkit";

const todoSlice = createSlice({
  name: "todos",
  initialState: [],
  reducers: {
    addTodo(state, action) {
      state.push({
        id: new Date().toISOString(),
        text: action.payload,
        completed: false,
      });
    },
    toggleComplete(state, action) {
      const toggledTodo = state.find((todo) => todo.id === action.payload);
      toggledTodo.completed = !toggledTodo.completed;
    },
    removeTodo(state, action) {
      return state.filter((todo) => todo.id !== action.payload);
    },
  },
});

export const { addTodo, toggleComplete, removeTodo } = todoSlice.actions;
export default todoSlice.reducer;
