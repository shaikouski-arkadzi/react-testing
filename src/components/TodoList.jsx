import { useSelector } from "react-redux";
import { selectorTodos } from "../store/selectors";
import TodoItem from "./TodoItem";

const TodoList = () => {
  const todos = useSelector(selectorTodos);

  return (
    <ul>
      {todos.map((el) => (
        <TodoItem key={el.id} {...el} />
      ))}
    </ul>
  );
};

export default TodoList;
