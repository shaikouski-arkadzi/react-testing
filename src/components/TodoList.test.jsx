import { describe, it } from "vitest";
import { render } from "@testing-library/react";
import * as reduxHooks from "react-redux";
import { Provider, useSelector } from "react-redux";
import TodoList from "../components/TodoList";
import store from "../store/store";

vi.mock("react-redux", async () => {
  const actual = await vi.importActual("react-redux");
  return {
    ...actual,
    useSelector: vi.fn(),
  };
});

describe("TodoList Component", () => {
  const todos = [
    { id: 1, text: "task 1", completed: false },
    { id: 2, text: "task 2", completed: true },
  ];

  it("matches the snapshot", () => {
    useSelector.mockReturnValue([]);

    const { asFragment } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should create TodoList with todo items", () => {
    useSelector.mockReturnValue(todos);

    const { asFragment } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should create TodoList with todo items alternative test", () => {
    const useSelectorSpy = vi
      .spyOn(reduxHooks, "useSelector")
      .mockReturnValue(todos);

    const { asFragment } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
    useSelectorSpy.mockRestore();
  });
});
