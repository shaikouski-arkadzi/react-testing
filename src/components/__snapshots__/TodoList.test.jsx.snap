// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList Component > matches the snapshot 1`] = `
<DocumentFragment>
  <ul />
</DocumentFragment>
`;

exports[`TodoList Component > should create TodoList with todo items 1`] = `
<DocumentFragment>
  <ul>
    <li>
      <input
        type="checkbox"
      />
      task 1
      <button>
        ×
      </button>
    </li>
    <li>
      <input
        checked=""
        type="checkbox"
      />
      task 2
      <button>
        ×
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`TodoList Component > should create TodoList with todo items alternative test 1`] = `
<DocumentFragment>
  <ul>
    <li>
      <input
        type="checkbox"
      />
      task 1
      <button>
        ×
      </button>
    </li>
    <li>
      <input
        checked=""
        type="checkbox"
      />
      task 2
      <button>
        ×
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
